### PR TITLE
proposal: [v2] Support fs.FS for assets

### LINKED ISSUE #1000: [v2] Support fs.FS for assets would allow some interesting scenarios
**Is your feature request related to a problem? Please describe.**
It would be great to not only have the possibility to embed the frontend but also e.g. use a `fs.FS` implementation for providing it.
This would allow some interesting scenarios like, allow overriding some files dynamically depending on the state of the app, or to have the frontend hosted on a webserver, e.g. as a squashfs. Having the frontend hosted somewhere would make updating the app, as long as there are only frontend changes, as easy as just publishing the new squashfs to the webserver. As soon as the user restarts the app, the updated frontend could be downloaded.

**Describe the solution you'd like**
Having the ability to provide a `fs.FS` implementation for the assets instead of `embed.FS`.

**Additional context**
I've created a draft implementation in my fork, and I'll provide it as a draft PR, where the proposal could be discussed. The only drawback currently is the auto-reload support in dev mode. This will only work if an `embed.FS` has been provided, in that case it is assumed, the provided `embed.FS` is serving the files from the `assetdir`.

### PR DESCRIPTION
**Pull request by stffabi**
Reloading changed asset files in dev mode will only work
if an embed.FS has been provided for the assets.

Proposal to fix #1000

### CODE REVIEW

File: v2/internal/frontend/assetserver/assetserver_common_dev.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+		} else {
+			// That's fine, the user is using another fs.FS and also has not set an assetdir
+		}
+	} else {
+		// We are in dev mode and an assetdir has been defined
+		if isEmbedFs {
+			// If the fs.FS is an embed.FS we assume it's serving the files from this directory, therefore replace the assets
+			// fs.FS with this directory.
+			// Are there any better ways to detect this? If we could somehow determine the compile time source path for the
+			// embed.FS this would be great. Currently there seems no way to get to this information.
```
[the remark this answers is not in the document]
Comment (stffabi) [reply v2/internal/frontend/assetserver/assetserver_common_dev.go:40]: I also thought about removing the assetdir flag during the implementation of the PR, and I would love it if we could remove it somehow.

It should work as long as the `embed.FS` variable has been defined on the main package, like it's done during a new project initialisation. The content to be included into the `embed.FS` is always relative to the package where the variable has been defined. If someone puts the variable into a sub package, let's say into the frontend package, we would only know the relative path `src` and we wouldn't know the relative path into the frontend package. In this case we would construct an absolute path of `project root` joined with `src`.

[the remark this answers is not in the document]
Comment (leaanthony) [reply v2/internal/frontend/assetserver/assetserver_common_dev.go:40]: Considering the frontend is *always* in projectroot/frontend, I think it would be easy to walk the embed.FS and see if it lines up, maybe check if the files exist in both places?

[the remark this answers is not in the document]
Comment (stffabi) [reply v2/internal/frontend/assetserver/assetserver_common_dev.go:40]: Yes, that's a good idea. We could determine the path to index.html in the embed.FS and see if `project root` joined with that relative path exists on the filesystem and also contains an index.html. And if it not lines up, we could log a warning. Should we leave the assetdir as a fallback or would you like to completely remove it?

[the remark this answers is not in the document]
Comment (leaanthony) [reply v2/internal/frontend/assetserver/assetserver_common_dev.go:40]: Initially fallback otherwise, if it doesn't line up you'll have problems

[the remark this answers is not in the document]
Comment (stffabi) [reply v2/internal/frontend/assetserver/assetserver_common_dev.go:40]: I'm going to update the PR tomorrow with these changes.

[the remark this answers is not in the document]
Comment (stffabi) [reply v2/internal/frontend/assetserver/assetserver_common_dev.go]: @leaanthony I've updated the PR, please let me know what you think.

File: v2/cmd/wails/internal/commands/dev/dev.go [hunk 2]
```diff
@@ -568,6 +601,35 @@ func doWatcherLoop(buildOptions *build.Options, debugBinaryProcess *process.Proc
 					debugBinaryProcess = newBinaryProcess
 				}
 			}
+			if len(changedPaths) != 0 {
+				if assetDir == "" {
+					resp, err := http.Get("http://localhost:34115/wails/assetdir")
+					if err != nil {
+						LogRed("Error during retrieving assetdir: %s", err.Error())
+					} else {
+						content, err := ioutil.ReadAll(resp.Body)
```
Comment (leaanthony) [inline v2/cmd/wails/internal/commands/dev/dev.go:610]: I think we should be using `io.ReadAll` since Go 1.16 :-)

Comment (leaanthony) [reply v2/cmd/wails/internal/commands/dev/dev.go:610]: https://tip.golang.org/doc/go1.16#ioutil

File: v2/cmd/wails/internal/commands/dev/dev.go [hunk 3]
```diff
@@ -83,8 +85,9 @@ func AddSubcommand(app *clir.Cli, w io.Writer) error {
 	flags := defaultDevFlags()
 	command.StringFlag("ldflags", "optional ldflags", &flags.ldflags)
 	command.StringFlag("compiler", "Use a different go compiler to build, eg go1.15beta1", &flags.compilerCommand)
-	command.StringFlag("assetdir", "Serve assets from the given directory", &flags.assetDir)
+	command.StringFlag("assetdir", "Serve assets from the given directory instead of using the provided asset FS", &flags.assetDir)
 	command.StringFlag("e", "Extensions to trigger rebuilds (comma separated) eg go", &flags.extensions)
+	command.StringFlag("reloaddirs", "Additional directories to trigger reloads (comma separated)", &flags.reloadDirs)
```
Comment (leaanthony) [inline v2/cmd/wails/internal/commands/dev/dev.go:90]: Why was this added? Not that I mind but I'd like to discuss it 👍

**Review by leaanthony — COMMENTED** [review]
Looking great! Just a couple of Qs.

File: v2/cmd/wails/internal/commands/dev/dev.go (hunk 2 above)
Comment (stffabi) [reply v2/cmd/wails/internal/commands/dev/dev.go:610]: Oh yes, you are right. I've not yet gotten used to it 😄 .

File: v2/cmd/wails/internal/commands/dev/dev.go (hunk 3 above)
Comment (stffabi) [reply v2/cmd/wails/internal/commands/dev/dev.go:90]: I've written a little bit more in the commit log about the use case. The main idea was to support the following scenario, that I currently have in our application.
I'm using some simple overlayFS implementation in a `fs.FS` interface, which is provided as the assets. The base FS comes from an `embed.FS` and the overlay content from a squashFS. During compilation the baseFS is either the `embed.FS` or a `os.DirFS` depending on the build tag. In combination with this new `reloadDirs` it's still possible to have reload triggering in dev mode for this scenario and because of the `os.DirFS` in dev mode, changes will be applied during reload.

Another idea I would see to add support for such a scenario could be the following. Having some specialed `wails.FS` interface, which might optionally be implemented to wrap an `fs.FS`. This new interface would bring some support for traveling up the tree of wrapped/overlay `fs.FS` interfaces and would bring support to replace an `fs.FS` implementation in the tree. In this case wails could replace an in-tree `embed.FS` with an `os.DirFS`. 
This would be rather complex and needs some detailed discussion, and the question is if it's worth the effort. Especially if the same could be achieved with the relatively small changeset of that commit. I also think using an `fs.FS` as assets is a more advanced topic, which might refine in the future and hopefully we will see more use cases and ideas how it's getting used. This might also bring more insight what user wants and help use define a more advanced support for "auto-reload" on `fs.FS` assets.

I would fully understand if you say this option heads into an undesired direction for wails.

**Comment by leaanthony** [comment]
Looks fine. Has it been tested on Windows + Mac?

**Comment by stffabi** [comment]
Yes, I've tested it on both Windows and Mac. My main development environment is Mac cross compiling to Windows, but I've also tested the dev mode on Windows.

**Comment by stffabi** [comment]
Thanks so much for merging all of my PRs. Awesome, this PR was the last one on my list to be upstreamed 🎉 . Therefore the cadence of PR's will decrease 😄 .

**Comment by leaanthony** [comment]
Windows being pure Go helps that for sure 😃 Cadence decrease? Oh no 😔
